fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownAggregateCollector.java

```diff
@@ -0,0 +1,521 @@
+        list.add(context);
+        allRewriteContext.put(context.origAggregator, list);
+        return null;
+    }
+
+    private boolean checkStatistics(ColumnRefSet groupBys, Statistics statistics) {
+        // check force push down flag
+        // flag 0: auto. 1: force push down. -1: don't push down. 2: push down medium. 3: push down high
+        if (sessionVariable.getCboPushDownAggregateMode() == 1) {
+            return true;
+        }
+
+        if (sessionVariable.getCboPushDownAggregateMode() == -1) {
+            return false;
+        }
+
+        List<ColumnStatistic> lower = Lists.newArrayList();
+        List<ColumnStatistic> medium = Lists.newArrayList();
+        List<ColumnStatistic> high = Lists.newArrayList();
+
+        List<ColumnStatistic>[] cards = new List[] {lower, medium, high};
+
+        groupBys.getStream().mapToObj(factory::getColumnRef)
+                .map(s -> ExpressionStatisticCalculator.calculate(s, statistics))
+                .forEach(s -> cards[groupByCardinality(s, statistics.getOutputRowCount())].add(s));
+
+        double lowerCartesian = lower.stream().map(ColumnStatistic::getDistinctValuesCount).reduce((a, b) -> a * b)
+                .orElse(Double.MAX_VALUE);
+
+        // pow(row_count/20, a half of lower column size)
+        double lowerUpper = Math.max(statistics.getOutputRowCount() / 20, 1);
+        lowerUpper = Math.pow(lowerUpper, Math.max(lower.size() / 2, 1));
+
+        // 1. white push down rules
+        // 1.1 only one lower/medium cardinality columns
+        if (high.isEmpty() && (lower.size() + medium.size()) == 1) {
+            return true;
+        }
+
+        // 1.2 the cartesian of all lower/count <= 1
+        // 1.3 the lower cardinality <= 3 and lowerCartesian < lowerUpper
+        // 1.4 follow medium cardinality flag
+        if (high.isEmpty() && medium.isEmpty()) {
+            if (lowerCartesian <= statistics.getOutputRowCount() || lower.size() <= 2) {
+                return true;
+            } else if (lower.size() <= 4 && lowerCartesian < lowerUpper) {
+                return true;
+            } else {
+                return sessionVariable.getCboPushDownAggregateMode() >= 2;
+            }
+        }
+
+        // 2. forbidden rules
+        // 2.1 high cardinality >= 2
+        // 2.2 medium cardinality > 2
+        // 2.3 high cardinality = 1 and medium cardinality > 0
+        if (high.size() >= 2 || medium.size() > 2 || (high.size() == 1 && !medium.isEmpty())) {
+            return false;
+        }
+
+        // 3. high cardinality < 2 and lower cardinality <= 3
+        if (high.size() == 1 && lower.size() <= 3) {
+            return sessionVariable.getCboPushDownAggregateMode() >= 3;
+        }
+
+        // 4. medium cardinality <= 2
+        if (lower.size() <= 3) {
+            if (sessionVariable.getCboPushDownAggregateMode() >= 2) {
+                return true;
+            }
+            return statistics.getOutputRowCount() >=
```

Comment:
if table data is very large, do we need to change this config?

Reply:
If the data of the table is larger, the aggregation effect should be better. This is a lower limit configuration